src/lib/kademlia/membership.ml

```diff
@@ -234,24 +238,27 @@ module Haskell_process = struct
           { failure_response= ref `Die
           ; process
           ; lock_path
-          ; already_waited= false }
+          ; terminated_ivar= Ivar.create () }
       with
       | Ok p ->
           (* If the Kademlia process dies, kill the parent daemon process. Fix
          * for #550 *)
           Deferred.bind (Process.wait p.process) ~f:(fun code ->
-              p.already_waited <- true ;
-              match (!(p.failure_response), code) with
-              | `Ignore, _ | _, Ok () ->
-                  return ()
-              | `Die, (Error _ as e) ->
-                  Logger.fatal logger ~module_:__MODULE__ ~location:__LOC__
-                    "Kademlia process died: $exit_or_signal"
-                    ~metadata:
-                      [ ( "exit_or_signal"
-                        , `String (Unix.Exit_or_signal.to_string_hum e) ) ] ;
-                  let%map () = Sys.remove lock_path in
-                  raise Child_died )
+              let%bind () =
+                match (!(p.failure_response), code) with
+                | `Ignore, _ | _, Ok () ->
+                    return ()
+                | `Die, (Error _ as e) ->
+                    Logger.fatal logger ~module_:__MODULE__ ~location:__LOC__
+                      "Kademlia process died: $exit_or_signal"
+                      ~metadata:
+                        [ ( "exit_or_signal"
+                          , `String (Unix.Exit_or_signal.to_string_hum e) ) ] ;
+                    raise Child_died
+              in
+              let%bind () = Sys.remove lock_path in
```

Comment:
Shouldn't the lock file be deleted before `raise`-ing above? Possibly the same for the `Ivar.fill` below, if hanging the caller of `kill` isn't okay.